reduce complexity of gdb waypt searchs from O(n^2) to O(n)

tsteven4 opened the pull request:
For the reader two parallel hashes are maintained with different
keys but identical values.  This is necessary because sometimes
we want to match the name and position, and other times we just want
to match the name.

tsteven4 committed:
use QHash for gdb waypoint searches.

This reduces the complexity from O(n^2) to O(n).

For the reader two parallel hashes are maintained with different
keys but identical values.  This is necessary because sometimes
we want to match the name and position, and other times we just want
to match the name.

tsteven4 committed:
Merge branch 'master' into gdbfindhash

robertlipe commented on gdb.cc:
```diff
@@ -70,19 +69,34 @@
+{
+  return
+    (lhs.shortname.compare(rhs.shortname, Qt::CaseInsensitive) == 0) &&
+    lhs.lat == rhs.lat &&
```
Aren't we more likely to have duplicates of names (especially names that are the same at the beginning) than by location? If so, consider changing the order of this test so that the numeric test is front of the short-circuit order.
Also, I know this is a behavioural change over The Way It's Always Been, but is a numeric comparison of double equality OK here? Im inclined to say that it is, as LHS and RHS would have the same incoming jitter when converted to doubles, but are they candidates for the abs(a-b) < kEpsilon trick?
...which may very well negate the first point.

OTOH, these numbers are hard to argue with and it's the way it's always been. {shrug}

tsteven4 replied:
yes, it mimics what has always been.  I'm going to treat this as out of scope, which was algorithm complexity improvements.

robertlipe commented on gdb.cc:
```diff
@@ -194,39 +208,45 @@ GdbFormat::gdb_fread_strlist() const
-    res = gdb_find_wayptq(&wayptq_in_hidden, wpt, exact);
+  Waypoint* res;
+  if (exact) {
+    res = gdb_find_wayptq(waypt_nameposn_in_hash, wpt);
```
This is much better than what I started and hadn't yet really come back to.

When pitching this, be sure that e knows that this is all possible because of the hundreds of hours we put into moving this into C++ a long time ago. Having ready access to indisputably "better" hashes than ones we created (I created) ourselves is just so much more powerful.

robertlipe commented on gdb.h:
```diff
@@ -63,6 +63,40 @@ class GdbFormat : public Format
+
+  class WptNamePosnKey;
+  using WptNamePosnHash = QHash<WptNamePosnKey, Waypoint*>;
+  class WptNamePosnKey {
```
Nice.

tsteven4 replied:
that one is ok, but the using of QHash::size_type was wrong!

tsteven4 replied:
CI rocks!

robertlipe reviewed (approved):
Nice. I started down a similar path but didn't get it finished. I like this a lot.

Please send those guys an invoice. :-)

tsteven4 committed:
workaround missing qHashMulti in Qt5.

and supply missing default for our qHash functions.

tsteven4 committed:
fix MSVC C2666 errors due to an erroneous assumption.

In Qt6 the return type of qHash, and the type of the seed,
happens to be the same as QHash::size_type.
However, this isn't true in Qt5.
